src/config.rs

```diff
@@ -181,3 +308,390 @@ pub fn from_env() -> Result<Session, Error> {
         Ok(session)
     }
 }
+
+#[cfg(test)]
+pub mod test {
+    use super::merge_mappings;
+    use super::*;
+    use env::set_current_dir;
+    use std::fs::File;
+    use std::io::Write;
+    use tempfile::tempdir;
+
+    #[test]
+    fn test_from_config() {
+        let dir = tempdir().unwrap();
+
+        let clouds_file_path = dir.path().join("clouds.yaml");
+        let mut clouds_file = File::create(&clouds_file_path).unwrap();
+        write!(
+            clouds_file,
+            r#"clouds:
+  cloud_name:
+    auth:
+      auth_url: http://url1
+      username: user1
+    profile: test_profile"#
+        )
+        .unwrap();
+
+        let clouds_public_file_path = dir.path().join("clouds-public.yaml");
+        let mut clouds_public_file = File::create(&clouds_public_file_path).unwrap();
+        write!(
+            clouds_public_file,
+            r#"public-clouds:
+  test_profile:
+    region_name: region1"#
+        )
+        .unwrap();
+
+        let secure_file_path = dir.path().join("secure.yaml");
+        let mut secure_file = File::create(&secure_file_path).unwrap();
+        write!(
+            secure_file,
+            r#"clouds:
+  cloud_name:
+    auth:
+      password: password1"#
+        )
+        .unwrap();
+
+        set_current_dir(&dir).unwrap();
+
+        assert!(from_config("cloud_name").is_ok());
```

Comment:
This fails in travis: https://travis-ci.org/github/dtantsur/rust-osauth/jobs/710216388. `assert(is_ok)` is hard to debug, could you change it (here and everywhere) to using `unwrap` or `expect`?

Reply:
I removed the is_ok() calls and I found why the tests are failing.
The tests fail intermittently because of std::env::set_current_dir which sets the current working directory for the whole process. It gets called multiple times in multiple threads during testing and there is a race condition.
I have verified this by running the tests with the --test-threads set to 1 (it never fails) and set to 10 (it fails intermittently).

I'm looking for a solution (maybe locking?) right now.

Reply:
I added a Mutex in the tests to prevent the race condition.